### PR TITLE
MH-13288 Don't create incomplete metadata fields

### PR DESCRIPTION
**Pull request by KatrinIhler**
Don't allow the creation of incomplete metadata fields where the functions valueToJSON and JSONToValue are not defined.

~~This PR includes PR #629.~~

### CODE REVIEW

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 1]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
-      case CONFIG_READ_ONLY_KEY:
-        this.readOnly = Boolean.valueOf(value);
-        break;
-      case CONFIG_REQUIRED_KEY:
-        this.required = Boolean.valueOf(value);
-        break;
-      case CONFIG_TYPE_KEY:
-        this.type = Type.valueOf(value.toUpperCase());
-        break;
+  public static MetadataField createMetadataField(Map<String,String> dublinCoreProperties) {
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: _dublinCoreProperties_ seems a suboptimal name to me: Configuration keys like "order", "delimiter", etc. are not related to Dublin Core at all.

Maybe something like metadataFieldConfig(uration) or similar?

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 2]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
+                        collectionID, order, namespace);
+      case TEXT_LONG:
+        return createTextLongMetadataField(inputID, outputID, label, readOnly, required, translatable, collection,
+                        collectionID, order, namespace);
+      case START_DATE:
+        return createTemporalStartDateMetadata(inputID, outputID, label, readOnly, required, pattern, order, namespace);
+      case START_TIME:
+        return createTemporalStartTimeMetadata(inputID, outputID, label, readOnly, required, pattern, order, namespace);
+      case ORDERED_TEXT:
+       return createOrderedTextMetadataField(inputID, outputID, label, readOnly, required, translatable, collection,
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Indentation

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 3]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
+                namespace);
+      case DURATION:
+        return createDurationMetadataField(inputID, outputID, label, readOnly, required, translatable,
+                collection, collectionID, order, namespace);
+      case ITERABLE_TEXT:
+        // Add an iterable text style field
+        return createIterableStringMetadataField(inputID, outputID, label, readOnly, required, translatable,
+                        collection, collectionID, delimiter, order, namespace);
+      case MIXED_TEXT:
+        // Add an iterable text style field
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Two times "iterable text style field"? I suggest to just remove those two comments as the method names already tell you which type of metadata field you add.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 4]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
+  public static MetadataField createMetadataField(String inputID, Opt<String> outputID, String label, boolean readOnly,
+          boolean required, Opt<Boolean> translatable, Type type, Opt<Map<String, String>> collection,
+          Opt<String> collectionID, Opt<Integer> order, Opt<String> namespace, Opt<String> delimiter, String pattern) {
+
+    switch (type) {
+      case BOOLEAN:
+        return createBooleanMetadata(inputID, outputID, label, readOnly, required, order, namespace);
+      case DATE:
+        return createDateMetadata(inputID, outputID, label, readOnly, required, pattern, order,
+                namespace);
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Hmm... would line 1088 really become longer than 120 characters without this line break?

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 5]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
+      case MIXED_TEXT:
+        // Add an iterable text style field
+        return createMixedIterableStringMetadataField(inputID, outputID, label, readOnly, required,
+                translatable, collection, collectionID, delimiter, order, namespace);
+      case LONG:
+        return createLongMetadataField(inputID, outputID, label, readOnly, required, translatable,
+                collection, collectionID, order, namespace);
+      case TEXT:
+        return createTextMetadataField(inputID, outputID, label, readOnly, required, translatable, collection,
+                        collectionID, order, namespace);
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Indentation not consistent within this method.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 6]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
-        this.required = Boolean.valueOf(value);
-        break;
-      case CONFIG_TYPE_KEY:
-        this.type = Type.valueOf(value.toUpperCase());
-        break;
+  public static MetadataField createMetadataField(Map<String,String> dublinCoreProperties) {
+    Opt<String> collectionID = dublinCoreProperties.containsKey(CONFIG_COLLECTION_ID_KEY)
+            ? Opt.some(dublinCoreProperties.get(CONFIG_COLLECTION_ID_KEY)) : Opt.none();
+    String pattern = dublinCoreProperties.containsKey(CONFIG_PATTERN_KEY)
+            ? dublinCoreProperties.get(CONFIG_PATTERN_KEY) : "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Maybe add this default value to the documentation of the metadata field _pattern_ on https://docs.opencast.org/develop/admin/configuration/metadata/

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java [hunk 7]
```diff
@@ -1041,60 +1038,84 @@ public static JObject mapToJSON(Map<String, String> map) {
...
+            Opt.none(), collectionID, order, namespace, delimiter, pattern);
+    metadataField.setListprovider(listprovider);
+    return metadataField;
+  }
+
+  public static MetadataField createMetadataField(String inputID, Opt<String> outputID, String label, boolean readOnly,
+          boolean required, Opt<Boolean> translatable, Type type, Opt<Map<String, String>> collection,
+          Opt<String> collectionID, Opt<Integer> order, Opt<String> namespace, Opt<String> delimiter, String pattern) {
+
+    switch (type) {
```
Comment (staubesv) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java:1003]: You would have the chance to add some sanity checks here, e.g. _inputID_ is a mandatory value, isn't it?

**Review by staubesv — CHANGES_REQUESTED** [review]
I've done a first round of review and have some minor suggestions for potential improvements (see comments).

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 2 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Fixed.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 5 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Fixed.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 3 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Fixed.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 4 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: Fixed.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 7 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java:1003]: Sanity checks are already done by the create methods called here and the constructor.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 1 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: You're right, other code calls it like that, but it doesn't make a lot of sense now that I think about it...

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java (hunk 6 above)
Comment (KatrinIhler) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/MetadataField.java]: I'm not at all sure whether that is consistent with other parts of the code, so I'm gonna refrain from that for now until all of this is a bit more sane...

**Comment by KatrinIhler** [comment]
@staubesv Updated!

**Comment by KatrinIhler** [comment]
@staubesv Can you take another look at this?

**Pull request merged**